### ISSUE TITLE
[Feature Request] (Develop Branch) Call unpuppet hooks after server crash.

### ISSUE THREAD

**Issue by kovitikus** [evennia/evennia#2829]
### Problem
When the server crashes, characters that were puppetted aren't properly cleaned up. The default behavior in Evennia is to gracefully unpuppet characters when the server goes down or when a player disconnects. A server crash is the only instance in which characters are not properly unpuppetted. This also leaves characters stuck on the grid when the server is started again, after a crash. Leaving the character in a potentially dangerous situation.

### Solution
In Evennia's `at_server_cold_start` hook, add logic to find characters that were puppetted during the crash and call their `at_pre_unpuppet`/`at_post_unpuppet` hooks.

This solution requires a way to track which characters were puppetted when the server crashed.

In this regard, a puppetted tag could be added to objects when they are puppetted, in the `at_post_puppet` hook. This tag can be subsequently removed in `at_post_unpuppet`.

The solution would look something like this.

```py
def at_server_cold_start():
    # Some galaxy brain will have to determine the best solution to grab all tagged objects.
    # Either with Evennia's tools for tagged object search or with a django query.
    for obj in get_all_puppetted_tagged_objects_here:
        if obj.location and not char.has_account:
            obj.at_pre_unpuppet()
            obj.at_post_unpuppet(None)
```
I'm uncertain if the `obj.location` check needs to happen. If an object tagged as `puppetted` is somehow already at `None` location, it won't harm it to have its `prelogout_location` set to `None` and its location set to `None` a second time.

Leaving out the `obj.location` check would ensure that objects tagged as `puppetted` get the tag cleaned up. After all, the `for` loop should only be iterating through a list of characters with the `puppetted` tag and we definitely want to clean up the tags!

Also, `obj.location` is already checked in the current `DefaultCharacter.at_post_unpuppet()` logic. So it won't attempt to set the `prelogout_location` or the new `location` if it already doesn't have a location.

Note that, even though there should be nobody connected to the game during the server startup, I still include the check for `has_account`. This is to avoid edge cases where some Chad might have some program that automatically attempts to reconnect to the game, login, and puppet a character through a script. Potentially re-puppetting a character before this logic is executed. For some reason, I really doubt this is possible, but I'm not certain and it's better to include the check for the edge case.

```py
class DefaultObject(ObjectDB, metaclass=TypeclassBase):
    def at_post_puppet(self, **kwargs):
        ...
        self.tags.add(key="puppetted")
    def at_post_unpuppet(self, account, session=None **kwargs):
        ...
        # The tag should be removed regardless if the object in question has a location.
        self.tags.remove(key="puppetted")
```

***

### Considerations

There are a few issues with this setup. First of all, the `DefaultObject` allows staff members to puppet it, but it does not contain the cleanup code present in `DefaultCharacter`'s implementation.

```py
# Current implementation on DefaultCharacter. Develop Branch.
class DefaultCharacter(DefaultObject):
    def at_post_unpuppet(self, account, session=None, **kwargs):
        """
        We stove away the character when the account goes ooc/logs off,
        otherwise the character object will remain in the room also
        after the account logged off ("headless", so to say).
        Args:
            account (Account): The account object that just disconnected
                from this object.
            session (Session): Session controlling the connection that
                just disconnected.
            **kwargs (dict): Arbitrary, optional arguments for users
                overriding the call (unused by default).
        """
        if not self.sessions.count():
            # only remove this char from grid if no sessions control it anymore.
            if self.location:

                def message(obj, from_obj):
                    obj.msg(
                        _("{name} has left the game.").format(name=self.get_display_name(obj)),
                        from_obj=from_obj,
                    )

                self.location.for_contents(message, exclude=[self], from_obj=self)
                self.db.prelogout_location = self.location
                self.location = None
```

One possible solution to this is to only set the `puppetted` tag on the `DefaultCharacter` typeclass. Alternatively, move the `self.db.prelogout_location = self.location` and the `self.location = None` lines up to its parent `DefaultObject` and at the end of the Character's hooks call `super()`.

***

Something else I noticed while going through this logic, that isn't exactly 100% related to the change, but does impact it, is the requirement to pass an `account` argument to `at_post_unpuppet`. Nothing exists in the `DefaultObject` post unpuppet hook and even the `DefaultCharacter` logic doesn't use the `account` argument!

It's worth considering making this `account` argument an optional argument, instead of required.

```py
def at_post_unpuppet(self, account=None, session=None, **kwargs):
    pass
```

This would make the `at_server_cold_start` logic cleaner.
```py
def at_server_cold_start():
    for obj in get_all_puppetted_tagged_objects_here:
        if obj.location and not obj.has_account:
            obj.at_pre_unpuppet()
            obj.at_post_unpuppet()
```

***
### Conclusion

This change brings the situation back in line with the rest of Evennia's puppetting logic. It also implements a new solution to finding currently puppetted characters via a tag that is only set when puppetted and cleaned up after the puppetting is over.

I'd also like to see if anybody else has ideas on how to use this new `puppetted` tag. Perhaps other puppetting cleanup or checks could be retrofitted to utilize the tag?

Do you have any issues that you think may arise from this change that I haven't considered?

**Comment by strikaco** [comment]
Only potential issue I see is that Tags may be unreliable in the event of failure. This might be more a job for a file-backed Django cache.

**Comment by kovitikus** [comment]
> Only potential issue I see is that Tags may be unreliable in the event of failure. This might be more a job for a file-backed Django cache.

I'm uncertain what you mean by the event of a failure. Tags are saved to the database when applied, so I'm not sure how it's unreliable. I don't anything about making a file-backed Django cache either.

**Comment by Griatch** [comment]
I don't think this only happens on a server crash, actually. I've seen a similar behavior over time on the demo server. I think it's rather an issue with what happens when the webclient window is closed abruptly rather than doing a clean log-off - it appears (maybe for certain browsers?) that the protocol is not noticing the logoff, and thus no unpuppeting hooks fire (them being removed from the grid when unpuppeting is the default side effect, not all games may want that to happen). So maybe the protocol would need some additional cleanup loop as well. 

Running the unpuppet methods on restart is a good idea (reminiscent to how we clean up non-persistent Scripts); I'm not sure what would be the better tracking method though; tags or something else. Overall, the link between puppets and accounts are deliberately weak in Evennia. :thinking:

**Comment by kovitikus** [comment]
It was suggested to me to check for an account tied to the character after the server restarts, rather than add tags. But, as you pointed out, that isn't exactly how Evennia wants to work.

I did some manual testing of a server crash by deliberately ending the process in Windows Task Manager.

This is what the character's information look like before the crash, while puppetted.
```
Name/key: Robert (#174)
Typeclass: Character (typeclasses.characters.Character)
Sessions: #2
Account: Tester (#2)
  Account Typeclass: Account (typeclasses.accounts.Account)
  Account Permissions: player
Location: City Square (#3)
Home: Limbo (#2)
Permissions: player
```

This is what the same chunk of information looks like after the server crashes.
```
Name/key: Robert (#174)
Typeclass: Character (typeclasses.characters.Character)
Location: City Square (#3)
Home: Limbo (#2)
Permissions: player
```

The reason I suggested using a tag is that it's very lightweight, it sticks around after a server crash since it is persistent, and the tag itself can quite easily be ignored by everything else (doesn't interfere with other systems).

The whole operation should be very lightweight itself, since very few objects will ever be puppetted.

***

My previous consideration about the unpuppetting cleanup missing in `evennia.objects.DefaultObject` versus the logic present in `evennia.objects.DefaultCharacter` is something I assume is present because you wouldn't want to move an object on or off the grid when puppetted or unpuppetted. Puppetting a rock object, for example, is probably something you'd do temporarily and it's a persistent object in the gameworld that you wouldn't want moved off the grid.

***

@Griatch I know that the `at_post_unpuppet` argument of `account` is a bit off topic here, but what do you think about turning that into an optional argument? Is there any reason it's a required argument? As I pointed out earlier, it's not even an argument that is used in the Evennia engine itself.

There's the possibility that changing it to optional would break stuff for people, but I figured now's the time to consider addressing this, since it's develop branch.

I understand why an account is a required parameter for `at_pre_puppet`, but it does not exist in either `at_post_puppet` or `at_pre_unpuppet`. It only reappears in `at_post_unpuppet`.

**Comment by InspectorCaracal** [comment]
> I don't think this only happens on a server crash, actually. I've seen a similar behavior over time on the demo server. I think it's rather an issue with what happens when the webclient window is closed abruptly rather than doing a clean log-off - it appears (maybe for certain browsers?) that the protocol is not noticing the logoff, and thus no unpuppeting hooks fire (them being removed from the grid when unpuppeting is the default side effect, not all games may want that to happen). So maybe the protocol would need some additional cleanup loop as well.

FYI that issue is probably related to/the same as #2734

**Comment by kovitikus** [comment]
https://github.com/evennia/evennia/blob/develop/evennia/server/server.py#L113

If `_server_maintenance` is run at server start, and then every 60 seconds afterwards, we could put my logic there and solve to problems at once. This way a stale character (tagged as puppetted, but `not obj.has_account`) would only remain on the grid for a maximum of 60 seconds before it gets cleaned up.

**Comment by InspectorCaracal** [comment]
My two cents:

I think the idea of a tag specific for characters being puppeted is probably the cleanest way to do it, since it would work equally well for any of the different multisession modes without any special handling and just using the existing hooks.

Also, whether or not the character gets moved back to `None` isn't really the issue - that's just the most visible symptom using default Evennia behavior. The actual problem is the unpuppet hooks not getting called, whatever the code in it might be. Whether a dev modified the default behavior or not, they're going to want it to run when a character is unpuppeted! So adding something to make sure that hook gets run would benefit everyone. And a maintenance check for link dead puppets after the idle season check sounds like a good solution to me.

**Comment by kovitikus** [comment]
> My two cents:
> 
> I think the idea of a tag specific for characters being puppeted is probably the cleanest way to do it, since it would work equally well for any of the different multisession modes without any special handling and just using the existing hooks.
> 
> Also, whether or not the character gets moved back to `None` isn't really the issue - that's just the most visible symptom using default Evennia behavior. The actual problem is the unpuppet hooks not getting called, whatever the code in it might be. Whether a dev modified the default behavior or not, they're going to want it to run when a character is unpuppeted! So adding something to make sure that hook gets run would benefit everyone. And a maintenance check for link dead puppets after the idle season check sounds like a good solution to me.

Yes, exactly! Thank you! 😊

**Comment by kovitikus** [comment]
This should work for the tag search itself.

```py
from evennia.objects.objects import DefaultCharacter

def _server_maintenance():
    ...
    for char in DefaultCharacter.objects.get_by_tag(key="puppetted"):
        if not char.has_account:
            char.at_pre_unpuppet()
            char.at_post_unpuppet()
```

**Comment by Griatch** [comment]
Thanks for the discussion and suggestions in this issue! I have now implemented a Tag similar to the one suggested here (but it's set at the Account level), and had the server maintenance task query for it. I also made the `account` arg of `at_post_unpuppet` optional (but for those that overrode this, we also send `None` here to minimize fuzz). 

As discussed in this thread, this will mean the link-dead puppet will at most hang around for 60 seconds. I also had the maintenance task add a `reason` to the kwargs of the hook, so one can output `Foo has left the game (connection lost).`

In the future, once SessionDBs are a thing, this will probably be possible to simplify, but this should work to resolve all link-dead issues for now.
